zcash_primitives/src/legacy/keys.rs

```diff
@@ -0,0 +1,253 @@
+            .derive_private_key(KeyIndex::hardened_from_normalize_index(account.0)?)
+            .map(AccountPrivKey)
+    }
+
+    pub fn from_extended_privkey(extprivkey: ExtendedPrivKey) -> Self {
+        AccountPrivKey(extprivkey)
+    }
+
+    pub fn to_account_pubkey(&self) -> AccountPubKey {
+        AccountPubKey(ExtendedPubKey::from_private_key(&self.0))
+    }
+
+    /// Derives the BIP-44 private spending key for the external (incoming payment) child path
+    /// `m/44'/<coin_type>'/<account>'/0/<child_index>`.
+    pub fn derive_external_secret_key(
+        &self,
+        child_index: u32,
+    ) -> Result<secp256k1::key::SecretKey, hdwallet::error::Error> {
+        self.0
+            .derive_private_key(KeyIndex::Normal(0))?
+            .derive_private_key(KeyIndex::Normal(child_index))
+            .map(|k| k.private_key)
+    }
+
+    /// Derives the BIP-44 private spending key for the internal (change) child path
+    /// `m/44'/<coin_type>'/<account>'/1/<child_index>`.
+    pub fn derive_internal_secret_key(
+        &self,
+        child_index: u32,
+    ) -> Result<secp256k1::key::SecretKey, hdwallet::error::Error> {
+        self.0
+            .derive_private_key(KeyIndex::Normal(1))?
+            .derive_private_key(KeyIndex::Normal(child_index))
+            .map(|k| k.private_key)
+    }
+}
+
+/// A type representing a BIP-44 public key at the account path level
+/// `m/44'/<coin_type>'/<account>'`.
+///
+/// This provides the necessary derivation capability for the for
+/// the transparent component of a unified full viewing key.
+#[derive(Clone, Debug)]
+pub struct AccountPubKey(ExtendedPubKey);
+
+impl AccountPubKey {
+    /// Derives the BIP-44 public key at the external "change level" path
+    /// `m/44'/<coin_type>'/<account>'/0`.
+    pub fn derive_external_ivk(&self) -> Result<ExternalIvk, hdwallet::error::Error> {
+        self.0
+            .derive_public_key(KeyIndex::Normal(0))
+            .map(ExternalIvk)
+    }
+
+    /// Derives the BIP-44 public key at the internal "change level" path
+    /// `m/44'/<coin_type>'/<account>'/1`.
+    pub fn derive_internal_ivk(&self) -> Result<InternalIvk, hdwallet::error::Error> {
+        self.0
+            .derive_public_key(KeyIndex::Normal(1))
+            .map(InternalIvk)
+    }
+
+    /// Derives the internal ovk and external ovk corresponding to this
+    /// transparent fvk. As specified in [ZIP 316][transparent-ovk].
+    ///
+    /// [transparent-ovk]: https://zips.z.cash/zip-0316#deriving-internal-keys
+    pub fn ovks_for_shielding(&self) -> (InternalOvk, ExternalOvk) {
+        let i_ovk = prf_expand_vec(
+            &self.0.chain_code,
+            &[&[0xd0], &self.0.public_key.serialize()],
+        );
+        let i_ovk = i_ovk.as_bytes();
+        let ovk_internal = InternalOvk(i_ovk[..32].try_into().unwrap());
+        let ovk_external = ExternalOvk(i_ovk[32..].try_into().unwrap());
```

Comment:
Bug: these are backwards.
```suggestion
        let ovk_external = ExternalOvk(i_ovk[..32].try_into().unwrap());
        let ovk_internal = InternalOvk(i_ovk[32..].try_into().unwrap());
```